### PR TITLE
Improve JS performance

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Along my random journey on the internet, I came upon an interesting fact when looking at some code: they were always setting the length of the data they were iterating on in a variable instead of just accessing it at every turn.

Based on this strange code, I decided to check why they were doing so and discovered that you could get up to 3% of performance improvements (which is big enough for me to consider it).

So to sum it up:

```js
var some_array = [];

for (var i = 0, len = some_array.length; i < len; ++i) {}
// is faster than:
for (var i = 0; i < some_array.length; ++i) {}
```

Of course, it only makes sense to use this when the data we're iterating on doesn't have its length updated while iterating.

The second commit is a little cleanup after I ran "eslint".

r? @jyn514

### CODE REVIEW

File: src/librustdoc/html/static/main.js [hunk 1]
```diff
@@ -649,8 +649,7 @@ function defocusSearchBar() {
          */
         function execQuery(query, searchWords, filterCrates) {
             function itemTypeFromName(typename) {
-                var length = itemTypes.length;
-                for (var i = 0; i < length; ++i) {
```
Comment (jyn514) [inline src/librustdoc/html/static/main.js:653]: Why does this change anything? The only difference is whether `len` is declared in the loop initializer, but the initializer only runs once, not each iteration of the loop.

Anyway, this is better written with the standard library:

```suggestion
var index = itemTypes.indexOf(typename);
if (index === -1) return NO_TYPE_FILTER;
return index;
```

File: src/librustdoc/html/static/main.js [hunk 2]
```diff
@@ -664,18 +663,15 @@ function defocusSearchBar() {
...
                 split = valLower.split("::");
 
-            var length = split.length;
-            for (var z = 0; z < length; ++z) {
+            for (var z = split.length - 1; z >= 0; --z) {
                 if (split[z] === "") {
                     split.splice(z, 1);
-                    z -= 1;
                 }
             }
```
Comment (jyn514) [inline src/librustdoc/html/static/main.js:671]:
```suggestion
split = split.filter(function(segment) { return segment !== ""; });
```

File: src/librustdoc/html/static/main.js [hunk 3]
```diff
@@ -782,17 +778,17 @@ function defocusSearchBar() {
...
                     // this validation does not make sense when searching by types
                     if (result.dontValidate) {
                         continue;
                     }
-                    var name = result.item.name.toLowerCase(),
-                        path = result.item.path.toLowerCase(),
-                        parent = result.item.parent;
+                    name = result.item.name.toLowerCase();
+                    path = result.item.path.toLowerCase();
+                    parent = result.item.parent;
```
Comment (jyn514) [inline src/librustdoc/html/static/main.js:792]: Does this really have a performance difference?

File: src/librustdoc/html/static/main.js [hunk 4]
```diff
@@ -868,19 +863,22 @@ function defocusSearchBar() {
...
+                                for (var y = 0; allFound === true && y < len; ++y) {
                                     allFound = false;
-                                    var firstGeneric = getObjectFromId(val.generics[y]).name;
-                                    for (x = 0; allFound === false && x < elems.length; ++x) {
+                                    firstGeneric = getObjectFromId(val.generics[y]).name;
+                                    elen = elems.length;
+                                    for (x = 0; allFound === false && x < elen; ++x) {
                                         allFound = getObjectFromId(elems[x]).name === firstGeneric;
                                     }
                                     if (allFound === true) {
```
Comment (jyn514) [inline src/librustdoc/html/static/main.js:885]:
```suggestion
var allFound = elems.every(function(id) { return getObjectFromId(id).name == firstGeneric; });
if (allFound === true) {
  elems.pop();
}
```
If you don't like the suggestion, you have a typo: `elen` -> `elems`.

File: src/librustdoc/html/static/main.js [hunk 5]
```diff
@@ -910,8 +908,7 @@ function defocusSearchBar() {
                 // Names didn't match so let's check if one of the generic types could.
                 if (literalSearch === true) {
                      if (obj.length > GENERICS_DATA && obj[GENERICS_DATA].length > 0) {
-                        var length = obj[GENERICS_DATA].length;
-                        for (x = 0; x < length; ++x) {
+                        for (x = 0, len = obj[GENERICS_DATA].length; x < len; ++x) {
                             if (obj[GENERICS_DATA][x] === val.name) {
                                 return true;
                             }
```
Comment (jyn514) [inline src/librustdoc/html/static/main.js:915]:
```suggestion
if (obj[GENERICS_DATA].some(function(name) { return name === val.name; })) {
  return true;
```

File: src/librustdoc/html/static/main.js [hunk 6]
```diff
@@ -927,8 +924,7 @@ function defocusSearchBar() {
                     lev_distance = Math.ceil((checkGenerics(obj, val) + lev_distance) / 2);
                 } else if (obj.length > GENERICS_DATA && obj[GENERICS_DATA].length > 0) {
                     // We can check if the type we're looking for is inside the generics!
-                    var olength = obj[GENERICS_DATA].length;
-                    for (x = 0; x < olength; ++x) {
+                    for (x = 0, len = obj[GENERICS_DATA].length; x < len; ++x) {
```
Comment (jyn514) [inline src/librustdoc/html/static/main.js:928]: Why this change?

File: src/librustdoc/html/static/main.js [hunk 7]
```diff
@@ -942,8 +938,7 @@ function defocusSearchBar() {
                 var lev_distance = MAX_LEV_DISTANCE + 1;
 
                 if (obj && obj.type && obj.type[INPUTS_DATA] && obj.type[INPUTS_DATA].length > 0) {
-                    var length = obj.type[INPUTS_DATA].length;
-                    for (var i = 0; i < length; i++) {
+                    for (var i = 0, len = obj.type[INPUTS_DATA].length; i < len; i++) {
```
Comment (jyn514) [inline src/librustdoc/html/static/main.js:942]: Why this change?

**Review by jyn514 — COMMENTED** [review]
How did you measure the performance change? I don't see why this would have anything to do with performace :confused:

**Comment by Lonami** [comment]
>  don't see why this would have anything to do with performace

Just a guess, but by having a local, the JIT compiler probably can do a better job with it. After all the original way always accesses the `.length` attribute of an arbitrary object which is probably more expensive than a local.

For the other changes where a `length` was declared before and is now also inside the `for`, I'd argue it makes it more consistent with the rest of these changes :P

**Comment by GuillaumeGomez** [comment]
@jyn514: I was running the performance comparison using https://jsbench.me/ for another project and came across this one. The 3% improvement is what came out after making the median out of the runs.

File: src/librustdoc/html/static/main.js (hunk 1 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/static/main.js:653]: It's maybe "better looking" but it's slower apparently (8% slower). I used this code to test:

```js
function foo(itemTypes, typename) {
	return itemTypes.indexOf(typename);
}
```

```js
function foo(itemTypes, typename) {
    for (var i = 0, len = itemTypes.length; i < len; ++i) {
	if (itemTypes[i] === typename) {
	     return i;
	}
    }
    return -1;
}
```

And then run the `foo` function in:

```js
var x = ["a", "akjsdfsdfs", "sdfqdfdqfqd", "kjsdbfqjkfqd", "qdkjfqdkjfqdfjkqsdfq", "qskjdfkqjsdb", "aizgeiazuedaz", "azlk,lcx,q", "qdsfns"];
for (var i = 0; i < 100000; ++i) {
	foo(x, "qdsfns");
}
```

(sill on https://jsbench.me/)

File: src/librustdoc/html/static/main.js (hunk 2 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/static/main.js:671]: That is a **huge** speed gain. (84% apparently). Great catch!

File: src/librustdoc/html/static/main.js (hunk 3 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/static/main.js:792]: No, it's more about cleanup. Recreating the variable at each iteration isn't "clean" (and invalid too, but js doesn't really care apparently).

File: src/librustdoc/html/static/main.js (hunk 4 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/static/main.js:885]: It's not a typo, it's meant for "elements length". But it's not that great so I'll update it a bit. Also, the code you suggested might be a worth a try, but I'll delay it for another PR until I fully get how it works.

File: src/librustdoc/html/static/main.js (hunk 5 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/static/main.js:915]: Actually, we could even use our own "onEach" function. It basically works the same after all.

File: src/librustdoc/html/static/main.js (hunk 6 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/static/main.js:928]: I removed a useless extra variable.

File: src/librustdoc/html/static/main.js (hunk 7 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/static/main.js:942]: More consistent with the rest.

File: src/librustdoc/html/static/main.js (hunk 1 above)
Comment (jyn514) [reply src/librustdoc/html/static/main.js:653]: Huh, weird. Well it's readable enough as-is so fine to keep it this way.

Comment (GuillaumeGomez) [reply src/librustdoc/html/static/main.js:653]: Welcome to the JS world. How do you like it so far? :p

File: src/librustdoc/html/static/main.js [hunk 8]
```diff
@@ -2830,25 +2821,27 @@ function defocusSearchBar() {
...
-            if (crates_text[i] === savedCrate) {
+            if (crate_text === savedCrate) {
                 elem.value = savedCrate;
             }
         }
         enableSearchInput();
     };
 
-    function buildHelperPopup() {
+    var buildHelperPopup = function() {
```
Comment (jyn514) [inline src/librustdoc/html/static/main.js:2844]: Why this change?

File: src/librustdoc/html/static/main.js [hunk 9]
```diff
@@ -569,7 +570,7 @@ function defocusSearchBar() {
                     len = rootPath.match(/\.\.\//g).length + 1;
 
                 for (i = 0; i < len; ++i) {
-                    match = url.match(/\/[^\/]*$/);
+                    match = url.match(/\/[^/]*$/);
```
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/librustdoc/html/static/main.js:573]: The linter complained about it so I removed it.

File: src/librustdoc/html/static/main.js (hunk 8 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/static/main.js:2844]: Because we call this function every time we want to show the help popup instead of creating the popup every time. However, since it only requires to be built once, we then set `buildHelperPopup` to an empty function so that calling it doesn't do anything anymore.

File: src/librustdoc/html/static/main.js (hunk 3 above)
Comment (frewsxcv) [reply src/librustdoc/html/static/main.js:788]: javascript variables defined with `var` are hoisted to the top of the function you're in, so declaring them earlier doesn't make a difference

Comment (frewsxcv) [reply src/librustdoc/html/static/main.js:788]: https://developer.mozilla.org/en-US/docs/Glossary/Hoisting

File: src/librustdoc/html/static/main.js [hunk 10]
```diff
@@ -927,8 +920,7 @@ function defocusSearchBar() {
                     lev_distance = Math.ceil((checkGenerics(obj, val) + lev_distance) / 2);
                 } else if (obj.length > GENERICS_DATA && obj[GENERICS_DATA].length > 0) {
                     // We can check if the type we're looking for is inside the generics!
-                    var olength = obj[GENERICS_DATA].length;
-                    for (x = 0; x < olength; ++x) {
+                    for (x = 0, len = obj[GENERICS_DATA].length; x < len; ++x) {
```
Comment (frewsxcv) [inline src/librustdoc/html/static/main.js:923]: where is `x` declared?

**Comment by jyn514** [comment]
I'll be honest, I am not comfortable making changes of this magnitude without tests. Maybe we could just keep the `filter` improvement (https://github.com/rust-lang/rust/pull/79052#discussion_r523768421) and leave the rest for later?

One way to verify this would be `use strict`, right? I don't think we have that on currently.

**Comment by GuillaumeGomez** [comment]
Actually, we use `"use strict";`. ;)

But if the amount of changes is problematic, I can make smaller PRs if you prefer?

**Comment by GuillaumeGomez** [comment]
Can be closed now!
